Libraries/Opc.Ua.Server/Subscription/SubscriptionManager.cs

```diff
@@ -1135,20 +1136,135 @@
             for (int ii = 0; ii < subscriptionIds.Count; ii++)
             {
                 TransferResult result = new TransferResult();
+                try
+                {
+                    // find subscription.
+                    Subscription subscription = null;
+                    lock (m_lock)
+                    {
+                        if (!m_subscriptions.TryGetValue(subscriptionIds[ii], out subscription))
+                        {
+                            result.StatusCode = StatusCodes.BadSubscriptionIdInvalid;
+                            results.Add(result);
+                            if ((context.DiagnosticsMask & DiagnosticsMasks.OperationAll) != 0)
+                            {
+                                diagnosticInfos.Add(null);
+                            }
+                            continue;
+                        }
+                    }
 
-                // find subscription.
-                Subscription subscription = null;
+                    // check if new and old sessions are different
+                    var oldSession = subscription.Session;
+                    if (oldSession != null && (oldSession.Id == null || oldSession.Id == NodeId.Null || oldSession.Id == context.Session.Id))
+                    {
+                        result.StatusCode = StatusCodes.BadNothingToDo;
+                        results.Add(result);
+                        if ((context.DiagnosticsMask & DiagnosticsMasks.OperationAll) != 0)
+                        {
+                            diagnosticInfos.Add(null);
+                        }
+                        continue;
+                    }
 
-                lock (m_lock)
-                {
-                    if (!m_subscriptions.TryGetValue(subscriptionIds[ii], out subscription))
+                    // The Server shall validate that the Client of that Session is operating on behalf of the same user
+                    // and that the potentially new Client supports the Profiles that are necessary for the Subscription.
+                    // --> Bad_UserAccessDenied
+                    if (oldSession != null && (!oldSession.IdentityToken.IsEqual(context.Session.IdentityToken)))
                     {
-                        result.StatusCode = StatusCodes.BadSubscriptionIdInvalid;
+                        result.StatusCode = StatusCodes.BadUserAccessDenied;
+                        results.Add(result);
+                        if ((context.DiagnosticsMask & DiagnosticsMasks.OperationAll) != 0)
+                        {
+                            diagnosticInfos.Add(null);
+                        }
+                        continue;
+                    }
+                    // --> Bad_InsufficientClientProfile
+                    if (!oldSession.CheckSecurityPolicyOfEndpoint(context.ChannelContext.EndpointDescription))
```

Comment:
need to check this helper...

Reply:
feel free ... 😃

Reply:
other feedback is all addressed, please let me know - what changes you need here

Reply:
Any update here?